### PR TITLE
Catelnuovo-Mumford regularity

### PR DESCRIPTION
**Pull request by wdecker**

### CODE REVIEW

File: src/Modules/UngradedModules.jl [hunk 1]
```diff
@@ -9089,3 +9089,48 @@ function has_monomials_on_all_axes(M::SubquoModule)
...
+1 -> x^2*e[1]
+2 -> y^2*e[1]
+3 -> z^2*e[1], F -> M
+e[1] -> e[1]
+Homogeneous module homomorphism)
+
+julia> regularity(M)
+3
+
+julia> minimal_betti_table(M);
```
Comment (lgoettgens) [inline src/Modules/UngradedModules.jl]: Why would you add the betti table to the doctest, if it doesn't get shown?

File: src/Modules/UngradedModules.jl [hunk 2]
```diff
@@ -9089,3 +9089,48 @@ function has_monomials_on_all_axes(M::SubquoModule)
...
+```
+"""
+function regularity(M::ModuleFP)
+ error("Not implemented for the given type")
+end
+
+function regularity(M::FreeMod)
+   return 0
```
Comment (lgoettgens) [inline src/Modules/UngradedModules.jl]: Shouldn't this check the base_ring to be graded as in the Subquo case?

File: src/Modules/UngradedModules.jl [hunk 3]
```diff
@@ -9089,3 +9089,48 @@ function has_monomials_on_all_axes(M::SubquoModule)
...
+end
+
+function regularity(M::SubquoModule)
+   @assert is_standard_graded( base_ring(M))
```
Comment (lgoettgens) [inline src/Modules/UngradedModules.jl]:
```suggestion
   @req is_standard_graded(base_ring(M)) "The base ring is not graded"
```
Adding a custom error message will be helpful for users.

File: src/Rings/mpoly-ideals.jl [hunk 4]
```diff
@@ -1647,3 +1647,32 @@ function grassmann_pluecker_ideal(ring::MPolyRing,
...
+# Examples
+```jldoctest
+julia> R, (w, x, y, z) = graded_polynomial_ring(QQ, ["w", "x", "y", "z"]);
+
+julia> I = ideal(R, [y^2*z − x^2*w, z^4 − x*w^3]);
+
+julia> regularity(I)
+6
+
+julia> minimal_betti_table(I);
```
Comment (lgoettgens) [inline src/Rings/mpoly-ideals.jl]: Same as above.

File: src/Rings/mpoly-ideals.jl [hunk 5]
```diff
@@ -1647,3 +1647,32 @@ function grassmann_pluecker_ideal(ring::MPolyRing,
...
+```
+"""
+function regularity(I::MPolyIdeal)
+   @assert is_standard_graded( base_ring(I))
```
Comment (lgoettgens) [inline src/Rings/mpoly-ideals.jl]:
```suggestion
   @req is_standard_graded(base_ring(I)) "The base ring is not graded"
```
Same as above.

**Review by lgoettgens — COMMENTED** [review]
Some minor remarks

File: src/Modules/UngradedModules.jl (hunk 1 above)
Comment (wdecker) [reply src/Modules/UngradedModules.jl]: There is a problem with the documenter in printing betti tables. The ; will be removed as soon as this problem is fixed.

File: src/Rings/mpoly-ideals.jl (hunk 4 above)
Comment (wdecker) [reply src/Rings/mpoly-ideals.jl]: as above

File: src/Modules/UngradedModules.jl (hunk 2 above)
Comment (wdecker) [reply src/Modules/UngradedModules.jl]: Yes, thanks, I will upload  a new file.

**Comment by micjoswig** [comment]
Our decision against submodules in OSCAR requires long names throughout.  Please rename regularity to CM_regularity or similar.

**Pull request merged**